subsys/bluetooth/services/dis.c

```diff
@@ -45,105 +47,231 @@ static struct dis_pnp dis_pnp_id = {
+				   ((uint64_t)CONFIG_BT_DIS_SYSTEM_ID_IDENTIFIER >> 32) & 0xFF,
+				   ((uint64_t)CONFIG_BT_DIS_SYSTEM_ID_IDENTIFIER >> 24) & 0xFF,
+				   ((uint64_t)CONFIG_BT_DIS_SYSTEM_ID_IDENTIFIER >> 16) & 0xFF,
+				   ((uint64_t)CONFIG_BT_DIS_SYSTEM_ID_IDENTIFIER >> 8) & 0xFF,
+				   ((uint64_t)CONFIG_BT_DIS_SYSTEM_ID_IDENTIFIER >> 0) & 0xFF};
+#endif
+
 #if defined(CONFIG_BT_DIS_SETTINGS)
 static uint8_t dis_model[CONFIG_BT_DIS_STR_MAX] = CONFIG_BT_DIS_MODEL;
 static uint8_t dis_manuf[CONFIG_BT_DIS_STR_MAX] = CONFIG_BT_DIS_MANUF;
 #if defined(CONFIG_BT_DIS_SERIAL_NUMBER)
-static uint8_t dis_serial_number[CONFIG_BT_DIS_STR_MAX] =
-	CONFIG_BT_DIS_SERIAL_NUMBER_STR;
+static uint8_t dis_serial_number[CONFIG_BT_DIS_STR_MAX] = CONFIG_BT_DIS_SERIAL_NUMBER_STR;
 #endif
 #if defined(CONFIG_BT_DIS_FW_REV)
-static uint8_t dis_fw_rev[CONFIG_BT_DIS_STR_MAX] =
-	CONFIG_BT_DIS_FW_REV_STR;
+static uint8_t dis_fw_rev[CONFIG_BT_DIS_STR_MAX] = CONFIG_BT_DIS_FW_REV_STR;
 #endif
 #if defined(CONFIG_BT_DIS_HW_REV)
-static uint8_t dis_hw_rev[CONFIG_BT_DIS_STR_MAX] =
-	CONFIG_BT_DIS_HW_REV_STR;
+static uint8_t dis_hw_rev[CONFIG_BT_DIS_STR_MAX] = CONFIG_BT_DIS_HW_REV_STR;
 #endif
 #if defined(CONFIG_BT_DIS_SW_REV)
-static uint8_t dis_sw_rev[CONFIG_BT_DIS_STR_MAX] =
-	CONFIG_BT_DIS_SW_REV_STR;
+static uint8_t dis_sw_rev[CONFIG_BT_DIS_STR_MAX] = CONFIG_BT_DIS_SW_REV_STR;
+#endif
+#if defined(CONFIG_BT_DIS_UDI)
+static uint8_t dis_udi_label[CONFIG_BT_DIS_STR_MAX] = CONFIG_BT_DIS_UDI_LABEL_STR;
+static uint8_t dis_udi_di[CONFIG_BT_DIS_STR_MAX] = CONFIG_BT_DIS_UDI_DI_STR;
+static uint8_t dis_udi_issuer[CONFIG_BT_DIS_STR_MAX] = CONFIG_BT_DIS_UDI_ISSUER_STR;
+static uint8_t dis_udi_authority[CONFIG_BT_DIS_STR_MAX] = CONFIG_BT_DIS_UDI_AUTHORITY_STR;
+#endif
+#if defined(CONFIG_BT_DIS_IEEE_RCDL)
+static uint8_t dis_ieee_rcdl[CONFIG_BT_DIS_STR_MAX] = CONFIG_BT_DIS_IEEE_RCDL_STR;
 #endif
 
-#define BT_DIS_MODEL_REF		dis_model
-#define BT_DIS_MANUF_REF		dis_manuf
-#define BT_DIS_SERIAL_NUMBER_STR_REF	dis_serial_number
-#define BT_DIS_FW_REV_STR_REF		dis_fw_rev
-#define BT_DIS_HW_REV_STR_REF		dis_hw_rev
-#define BT_DIS_SW_REV_STR_REF		dis_sw_rev
+#define BT_DIS_MODEL_REF             dis_model
+#define BT_DIS_MANUF_REF             dis_manuf
+#define BT_DIS_SERIAL_NUMBER_STR_REF dis_serial_number
+#define BT_DIS_FW_REV_STR_REF        dis_fw_rev
+#define BT_DIS_HW_REV_STR_REF        dis_hw_rev
+#define BT_DIS_SW_REV_STR_REF        dis_sw_rev
+#define BT_DIS_UDI_LABEL_STR_REF     dis_udi_label
+#define BT_DIS_UDI_DI_STR_REF        dis_udi_di
+#define BT_DIS_UDI_ISSUER_STR_REF    dis_udi_issuer
+#define BT_DIS_UDI_AUTHORITY_STR_REF dis_udi_authority
+#define BT_DIS_IEEE_RCDL_STR_REF     dis_ieee_rcdl
+
+/*
+ * When assigning too long string literals to the arrays,
+ * the literals may be truncated, removing the null terminator.
+ * Using strnlen to avoid sending data outside the array.
+ */
+#define BT_DIS_STRLEN(x) strnlen(x, CONFIG_BT_DIS_STR_MAX)
```

Comment:
Looks like `CONFIG_BT_DIS_STR_MAX` also includes the null terminator, right?

If so, shouldn't this be
```suggestion
#define BT_DIS_STRLEN(x) strnlen(x, CONFIG_BT_DIS_STR_MAX - 1)
```

as we don't want this to return values `>= CONFIG_BT_DIS_STR_MAX`

Reply:
In the Kconfig, it says that `BT_DIS_STR_MAX` is:
> Bluetooth DIS string storage size. Storage can be up to 248 bytes long (excluding NULL termination).

If we use `CONFIG_BT_DIS_STR_MAX` in `strnlen` and for the array sizes, this allows the strings to be `CONFIG_BT_DIS_STR_MAX` characters long.
`strnlen` will never return a value larger than `CONFIG_BT_DIS_STR_MAX`

Reply:
But if `CONFIG_BT_DIS_STR_MAX` doesn't include the NULL terminator, then all arrays that are using this as the size need to be `+ 1`, right?

Reply:
That was what I initially thought as well, but if you assign a too long string, it will be truncated at the array length. That means that we will not be guaranteed to have a null-terminator unless we add it manually. We do not need it however, `strnlen` does the job. The BLE methods just take a length, so they do also not need a null-terminator to function.